### PR TITLE
TR-DOS filename related fixes (SAVETRD, SAVEHOB)

### PR DESCRIPTION
**Pull request by Liniya**
This mostly concerns an old bug where dots weren't properly handled in TR-DOS filenames. Previously, sjasmplus used to assume that the first dot encountered in the TR-DOS filename is followed by an extension (1 character), so a filename like 'abc.def.C' would be truncated to 'abc' with the type 'd' in the TRD image. The fix was to change the extension test method, always checking only the second-last character in the TR-DOS filename for the dot, for filenames with length > 2. As well, it will now produce an error if the input TR-DOS filename is too long.

It was complicated by the fact that it had an additional hidden 'feature' of sorts: if two characters after the extension in the input TR-DOS filename were not zero, they were treated as an uint16 and the start address field of the file was set to this value. However, I couldn't find instances of this feature being used inside sjasmplus, and there doesn't appear to be a sensible way to use it in ASM sources either (outside of doing some obscure LUA string manipulations perhaps). Since it also prevented reliably checking TR-DOS extensions, I have removed it. Personally, I think the best course of action to add this sort of functionality is with adding a regular (optional) comma-separated parameter to the save functions.

### CODE REVIEW

**Comment by Liniya** [comment]
Note about the checks: it is failing the test 'sjasmplus_regressions\custom\trd.asm'. This I think is because the test uses incorrect TR-DOS filenames like 'label1.txt'. A TR-DOS filename is max.8 characters, with a single-character extension.

**Comment by ped7g** [comment]
The test is copied from the other sjasmplus: https://github.com/sjasmplus/sjasmplus/blob/master/regression/custom/trd.asm ... feel free to fix it to be more correct, or I will try myself when I will get to your pull request (or whoever will). But from a quick look this looks very good, thank you. :) This will surely help.

**Comment by Liniya** [comment]
Actually, I'm not sure if this situation warrants an error. For now, I've changed it so that it only uses the first character of the provided extension, discarding the rest (no error is given), which is a middle ground between now and before. As such, I didn't change the test, although marked it as using incorrect extensions.

File: sjasm/io_trd.cpp [hunk 1]
```diff
@@ -167,6 +167,7 @@ int TRD_AddFile(char* fname, char* fhobname, int start, int length, int autostar
 	}
 
 	//header of file
+/*
```
Comment (ped7g) [inline sjasm/io_trd.cpp]: Can you just delete the old code? It just clutters current version, there's no need for it (is archived in git), and it's making more difficult for git to find true changes to the code, so harder to review it.

File: sjasm/io_trd.cpp [hunk 2]
```diff
@@ -195,8 +196,32 @@ int TRD_AddFile(char* fname, char* fhobname, int start, int length, int autostar
...
+
+	if (hdr[8] == 'B')	{
+		hdr[0x09] = (unsigned char)(length & 0xff);
+		hdr[0x0a] = (unsigned char)(length >> 8);
+	} else	{
+		hdr[0x09] = (unsigned char)(start & 0xff);
+		hdr[0x0a] = (unsigned char)(start >> 8);
 	}
 
+
```
Comment (ped7g) [inline sjasm/io_trd.cpp]: try to avoid pure whitespace changes of old source (or put them into separate commit focused on restyling old code, to make them separate from code changes)

File: sjasm/sjio.cpp [hunk 3]
```diff
@@ -960,6 +960,7 @@ int SaveBinary(char* fname, int start, int length) {
 int SaveHobeta(char* fname, char* fhobname, int start, int length) {
 	unsigned char header[0x11];
 	int i;
+/*
```
Comment (ped7g) [inline sjasm/sjio.cpp]: please just delete the old block

File: sjasm/io_trd.cpp [hunk 4]
```diff
@@ -195,8 +196,32 @@ int TRD_AddFile(char* fname, char* fhobname, int start, int length, int autostar
...
+	if (i > 1)
+	{
+		char *ext = strrchr(fhobname, '.');
+		if (ext && ext[1])
+		{
+			hdr[8] = ext[1];
+			*ext = '\0';
+			i = strlen(fhobname);
+		}
+	}
```
Comment (ped7g) [inline sjasm/io_trd.cpp]: Maybe warning about truncating extension and/or name can be issued here? If that makes sense... like
`if (8 < i || (ext && ext[1] && ext[2])) { Warning("Invalid filename in trdos, will get truncated", fhobname); }` (and needs no change to fhobname, will add comment above how to achieve that)

File: sjasm/io_trd.cpp [hunk 5]
```diff
@@ -195,8 +196,32 @@ int TRD_AddFile(char* fname, char* fhobname, int start, int length, int autostar
...
+	memset(hdr,' ',9);
+	i = strlen(fhobname);
+	if (i > 1)
+	{
+		char *ext = strrchr(fhobname, '.');
+		if (ext && ext[1])
+		{
+			hdr[8] = ext[1];
+			*ext = '\0';
+			i = strlen(fhobname);
```
Comment (ped7g) [inline sjasm/io_trd.cpp]: you can avoid `fhobname` modification and extra `strlen` by `i = ext-fhobname;`

File: sjasm/io_trd.cpp (hunk 4 above)
Comment (Liniya) [reply sjasm/io_trd.cpp]: Strictly speaking, you are probably right in that there should be at least some indication about it, and this was also my initial idea. However, I think this is ideally decided/changed together with other similar save functions like SAVETAP (it doesn't make sense if one function produces warnings and the other does not), which is outside of the scope of this fix.
